### ISSUE TITLE
test stuck at `eyes.open` forever

### ISSUE THREAD

**Issue by Bobgy** [applitools/eyes.cypress#12]
I tried playing around with applitools cypress SDK. At first, it's all good, but after I implement a few tests and trying to run all tests in a row. It randomly stuck at different tests on the "cy.eyesOpen" statement. My arguments passed aren't special at all, just `appName`, `testName` and `browser`.

So I don't quite understand what went wrong.

**Comment by amitzur** [comment]
Hi @Bobgy, could you run the test with logs enabled and send us the result?

You can enable logs by either defining the environment variable `APPLITOOLS_SHOW_LOGS=true` or by specifying in the `applitools.config.js` file the property `showLogs: true`.

Thanks!

**Comment by Bobgy** [comment]
Hi @amitzur, I sent the log.

**Comment by amitzur** [comment]
Thanks, we're checking and will update both via email and here.

**Comment by amitzur** [comment]
This issue was reproduced, and shows an interesting case, so it'll be helpful to describe here:
Due to a failure during one of the Cypress tests, the Applitools test was not called (since the `cy.eyesClose` command was not reached).
Now, by default we don't run tests in parallel for free accounts (same as specifying `concurrency===1` in the `applitools.config.js`), so if a test doesn't close, the following test will not open and the whole run will be stuck, which is the symptom of this issue.

There is a way to alter this behavior, and that is by placing the open/close commands inside beforeEach/afterEach respectively.
There's a section in our docs explaining how:
https://www.npmjs.com/package/@applitools/eyes.cypress#best-practice-for-using-the-sdk
This way, close will be called even if a test failed on something.

Here's an example for a test run with an error that will work properly:
```
describe('Hello world', () => {
  beforEach(() => {
    cy.eyesOpen({appName: 'Hello World!'});
  });
 
  afterEach(() => {
    cy.eyesClose();
  });
 
  it('test 1', () => {
    cy.visit('https://applitools.com/helloworld');
    cy.eyesCheckWindow('Main Page').then(() => {
      throw new Error('this test has failed');
    });
  });

  it('test 2', () => {
    cy.get('button').click();
    cy.eyesCheckWindow('Click!');
  });
});
```

**Comment by Bobgy** [comment]
@amitzur I think it's worth updating Cypress tutorial too. https://applitools.com/tutorials/cypress.html#quick-start

**Comment by Bobgy** [comment]
also, I think this is still considered bad practice for cypress: https://docs.cypress.io/guides/references/best-practices.html#Using-after-or-afterEach-hooks

**Comment by Bobgy** [comment]
and even worse, I'm not sure if I am missing something.
I followed best practice examples and got the following error during eyesOpen
```
Error: IllegalArgument: testName is null or undefined
```

Because each `eyesOpen` command should have a different test name, I cannot tell how `beforeEach` hook can work with it.
On the other hand. I couldn't put `eyesClose` to afterEach hook without putting `eyesOpen` to beforeEach hook, because not all my tests need visual regression testing. And those who don't will call `eyesClose` without calling `eyesOpen`, which is an error too.

**Comment by amitzur** [comment]
> @amitzur I think it's worth updating Cypress tutorial too. https://applitools.com/tutorials/cypress.html#quick-start

definitely.

> also, I think this is still considered bad practice for cypress: https://docs.cypress.io/guides/references/best-practices.html#Using-after-or-afterEach-hooks

Cypress recommends using `before` and `beforeEach` rather than `after` and `afterEach` for *resetting state*. This is not the same as calling `cy.eyesOpen` and `cy.eyesClose`, which means starting a transaction and ending it.
Cypress' recommendation also refers more to what you do to your application, rather than ways to use external SDK's on top of your tests.
The only argument that stands is if the user refreshes Cypress at the middle of the test. The outcome of this is that you _might_ have Applitools tests that remain in `running` state. They can be aborted through the UI of the test manager. 
I agree with Cypress' docs that it's a best practice to perform cleanups before tests for anything that might have accumulated in previous runs.

> `Error: IllegalArgument: testName is null or undefined`

This shouldn't happen. Would you mind sharing your script and config file?

> Because each eyesOpen command should have a different test name, I cannot tell how beforeEach hook can work with it.

It's possible to get the currently running test name within the `beforeEach` block by accessing `this.currentTest`. This is exactly how we do it.

> On the other hand. I couldn't put eyesClose to afterEach hook without putting eyesOpen to beforeEach hook, because not all my tests need visual regression testing. And those who don't will call eyesClose without calling eyesOpen, which is an error too.

This is why we recommend (or rather, require) symmetry in calling `eyesOpen` in `beforeEach` and `eyesClose` in `afterEach`.

**Comment by Bobgy** [comment]
Hi @amitzur , sorry for the late reply. I'm back from vacation.
> Error: IllegalArgument: testName is null or undefined
Thank you for the info, I figured out the issue is something I did wrongly by myself.

I tried your approach of putting `cy.eyesOpen` and `cy.eyesClose` in beforeEach and afterEach hooks. Although `cy.eyesClose` was called successfully, some tests were still stuck. Then I upgraded `@applitools/eyes.cypress` to latest `3.3.3`, after that, the issue is gone.

Then I hit yet another issue, saw the error message so I figured out that I was passing `tag` as null to eyes.checkWindow, which could also let tests stuck forever.

After fixing the third issue, my tests are running well.

**Comment by Bobgy** [comment]
Thank you for the help!

**Comment by Bobgy** [comment]
@amitzur after setting it up with all my tests like this, I noticed another error.
```
Error:       Error: Eyes.Cypress detected diffs or errors during execution of visual tests:
       [32mPassed - 1 tests[39m
         [32m✓[39m [0mpreferences filter should work [375x667][0m
       [31mDiffs detected - 0 tests[39m
       [31mErrors - 1 tests[39m
         [31m✖[39m [0mError: stopped transaction before it started[0m

Because this error occurred during a 'after all' hook we are skipping all of the remaining tests.
```

After spending some time locating the issue, I think it's caused by colocating two tests in one `describe`, while one does visual testing, the other doesn't.

example:
```
beforeEach(() => cy.eyesOpen({ ... }));
afterEach(() => cy.eyesClose({ ... }));
describe('test', () => {
  it('do visual testing', () => {
    cy.eyesCheckWindow()
  }
  it('doesn't do visual testing', () => {
    ...
  }
}
```

Any ideas how I can fix it?

**Comment by amitzur** [comment]
Good catch!
Indeed, we didn't handle that situation well. But it's fixed now in v3.3.8 so please update, and thanks for reporting!

**Comment by Bobgy** [comment]
@amitzur awesome, I'll try

**Comment by Bobgy** [comment]
just FYI, I tested this again and it works as expected
